parlai/tasks/ccpe/agents.py

```diff
@@ -3,150 +3,106 @@
 # This source code is licensed under the MIT license found in the
 # LICENSE file in the root directory of this source tree.
 
-from parlai.core.teachers import FixedDialogTeacher
+from parlai.core.teachers import DialogTeacher
 from parlai.utils.io import PathManager
+from parlai.utils.data import DatatypeHelper
 from .build import build
 import os
 import json
 
 
-class CCPEAllTeacher(FixedDialogTeacher):
+class _Abstract(DialogTeacher):
     def __init__(self, opt, shared=None):
-        # store datatype
-        super().__init__(opt, shared)
-
-        dt = opt['datatype'].split(':')[0]
-        if dt != 'train':
-            raise RuntimeError('Not valid datatype (only train).')
-
-        if shared:
-            self.data = shared['data']
-        else:
+        opt['datafile'] = DatatypeHelper.fold(opt['datatype'])
+        self.dpath = os.path.join(opt["datapath"], "CCPE")
+        if shared is None:
             build(opt)
-            self._setup_data()
-
-        self.reset()
-
-    def num_episodes(self):
-        return len(self.data)
-
-    def num_examples(self):
-        return sum([len(x) for x in self.data])
-
-    def _setup_data(self):
+        super().__init__(opt, shared)
 
+    def _get_turns_of_speaker(self, utts, idx, speaker):
```

Comment:
what exactly is this function doing?

Reply:
Compressing all contiguous turns of a given speaker (all of a user or all of an Assistant or all of a User) in the original json into one. Adding in newlines is pretty similar to what the previous version of this was doing.. just that it's abstracted out here. (Can add a comment to make this more clear.)